packages/contracts-bedrock/test/kontrol/proofs/utils/KontrolUtils.sol

```diff
@@ -0,0 +1,154 @@
+        _arr[0] = ghostBytes0;
+        _arr[1] = ghostBytes1;
+        _arr[2] = ghostBytes2;
+        _arr[3] = ghostBytes3;
+        _arr[4] = ghostBytes4;
+        _arr[5] = ghostBytes5;
+        _arr[6] = ghostBytes6;
+        _arr[7] = ghostBytes7;
+        _arr[8] = ghostBytes8;
+        _arr[9] = ghostBytes9;
+    }
+}
+
+/// @notice tests inheriting this contract cannot be run with forge
+abstract contract KontrolUtils is KontrolCheats {
+    /// @dev we only care about the vm signature
+    // Cheat code address, 0x7109709ECfa91a80626fF3989D68f67F5b1DD12D.
+    address private constant VM_ADDRESS = address(uint160(uint256(keccak256("hevm cheat code"))));
+    Vm private constant vm = Vm(VM_ADDRESS);
+
+    function createWithdrawalTransaction(
+        uint256 _tx0,
+        address _tx1,
+        address _tx2,
+        uint256 _tx3,
+        uint256 _tx4,
+        bytes memory _tx5
+    )
+        internal
+        pure
+        returns (Types.WithdrawalTransaction memory _tx)
+    {
+        _tx = Types.WithdrawalTransaction(_tx0, _tx1, _tx2, _tx3, _tx4, _tx5);
+    }
+
+    function freshBytesArray(uint256 symbolicArrayLength) public returns (bytes[] memory symbolicArray) {
+        symbolicArray = new bytes[](symbolicArrayLength);
+
+        for (uint256 i = 0; i < symbolicArray.length; ++i) {
+            symbolicArray[i] = abi.encodePacked(kevm.freshUInt(32));
+        }
+    }
+
+    /// @dev Returns a symbolic bytes32
+    function freshBytes32() public returns (bytes32) {
+        return bytes32(kevm.freshUInt(32));
+    }
+
+    /// @dev Returns a symbolic adress
+    function freshAdress() public returns (address) {
+        return address(uint160(kevm.freshUInt(20)));
+    }
+
+    /// @dev Creates a fresh bytes with length greater than 31
+    /// @param bytesLength: Length of the fresh bytes. Should be concrete
+    function freshBigBytes(uint256 bytesLength) internal returns (bytes memory sBytes) {
+        require(bytesLength >= 32, "Small bytes");
+
+        uint256 bytesSlotValue;
+        unchecked {
+            bytesSlotValue = bytesLength * 2 + 1;
+        }
+
+        /* Deploy ghost contract */
+        GhostBytes ghostBytes = new GhostBytes();
+
+        /* Make the storage of the ghost contract symbolic */
+        kevm.symbolicStorage(address(ghostBytes));
+
+        /* Load the size encoding into the first slot of ghostBytes*/
+        vm.store(address(ghostBytes), bytes32(uint256(0)), bytes32(bytesSlotValue));
+
+        sBytes = ghostBytes.ghostBytes();
+    }
+
+    /// @dev Creates a bounded symbolic bytes[] memory representing a withdrawal proof
+    /// Each element is 17 * 32 = 544 bytes long, plus ~10% margin for RLP encoding: each element is 600 bytes
+    /// The length of the array to 10 or fewer elements
+    function freshWithdrawalProof() public returns (bytes[] memory withdrawalProof) {
+        /* Assume arrayLength = 2 for faster proof speeds */
+        uint256 arrayLength = 10;
```

Comment:
Is the proof / this method assuming the proof length is *exactly* 10 elements, or that the max length is 10 elements?

Reply:
The proof is assuming that the length is _exactly_ 10 elements. I would wait until branch parallelization is implemented since assuming a bound of 10 would imply branching on each possible value. If there's a better temporal fixed value other than 10, let me know and I'll change it!

Reply:
How much of a speedup would we get by changing this from exactly 10 to exactly 6 or 7? In practice right now I haven't found one longer than 6 elements—using 10 here was intended to be conservative to future proof as the proof length grows over time. But I'd be ok with using 6 or 7 for now, and changing to a bound of 10 once branch parallelization is implemented.

For reference, I manually looked at the input data for `proveWithdrawalTransactions` here: https://etherscan.io/txs?a=0xbEb5Fc579115071764c7423A4f12eDde41f106Ed

Reply:
Right now, some hours of computation, I believe. But this won't resolve the core problem I think, since the execution time will still be too high (currently under investigation). Regardless, it's great to know we can default to 6 or 7! Then bump it to a bound of 10 once branch parallelization is in place